Add process.dlopenFlags

ezequielgarcia opened the pull request:
This PR is another attempt at support dlopen flags. A previous try was made in https://github.com/nodejs/node/pull/4105, but it never got accepted.

In particular, it's common to require a node addon to be loaded using `dlopen(flag=RTLD_GLOBAL)`. For instance, this is required to implement a libao binding, due to libao's plugins being shared objects that need symbols exposed in a common shared object (see [here](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=656967)).

It's interesting to mention that this feature is being requested since at least 2010:

https://github.com/nodejs/node-v0.x-archive/issues/436
https://groups.google.com/forum/#!topic/nodejs/L2wNb_azzyU

I really hope we can tackle it this time, so please provide any feedback to make this feature acceptable.

The code is a little bit ugly, open-coding the `dlopen` call instead of abstracting it behind libuv. My initial attempt was through libuv, but it got rejected: https://github.com/libuv/libuv/pull/1331.

This is inspired by Python `sys.{get,set}dlopenflags`. Python docs [here](https://docs.python.org/2/library/sys.html#sys.setdlopenflags) and [here](https://docs.python.org/2/library/sys.html#sys.getdlopenflags).

##### Checklist

- [x] `make -j4 test` (UNIX) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
src, process

addaleax commented on doc/api/process.md:
```diff
@@ -683,6 +683,22 @@ process.on('warning', (warning) => {
 
+## process.dlopenFlags
+<!-- YAML
+added: v8.0.0
```
Please use `added: REPLACEME` :)

mscdex commented on doc/api/os.md:
```diff
@@ -1160,6 +1160,38 @@ The following error codes are specific to the Windows operating system:
+### dlopen Constants
+
+The following signal constants are exported by `os.constants.dlopen`.
+See `man 3 dlopen` for detailed informatio.
```
s/informatio/information/

ezequielgarcia replied:
done

ezequielgarcia replied:
OK.

mscdex commented on doc/api/os.md:
```diff
@@ -1160,6 +1160,38 @@ The following error codes are specific to the Windows operating system:
+    <th>Constant</th>
+  </tr>
+  <tr>
+    <td><code>RTLD_LAZY</code></td>
```
Missing descriptions for these constants?

ezequielgarcia replied:
I thought the descriptions were too detailed and complex, and was expecting the reader could refer to the manual instead? It's a 1:1 mapping of the `dlopen` flags, so there's not any secret sauce.

jasnell replied:
If the idea is to refer the user to a manual, then having these as links to that manual would be ideal :-) ... still, a very brief one sentence description would be preferred (I know we're fairly inconsistent with that in the docs and we need to get better at it)

ezequielgarcia replied:
@jasnell I've added a line with some description. I agree it looks better now.

mscdex commented on src/node_constants.cc:
```diff
@@ -1287,6 +1321,10 @@ void DefineConstants(v8::Isolate* isolate, Local<Object> target) {
 
+  Local<Object> dlopen_constants = Object::New(isolate);
+  CHECK(dlopen_constants->SetPrototype(env->context(),
+                                     Null(env->isolate())).FromJust());
```
This line should be indented to align with the first argument to `SetPrototype()`.

ezequielgarcia replied:
OK.

mscdex commented on src/node.cc:
```diff
@@ -2502,18 +2506,44 @@ struct node_module* get_linked_module(const char* name) {
+
+#ifdef __POSIX__
+  // If no flag is passed, fallback to default
+  if (args.Length() < 3 || args[2]->IsUndefined() ||
```
What about falling back to the old behavior if it's not an integer instead of just `undefined` or `null`?

ezequielgarcia replied:
Sure, that works.

ezequielgarcia replied:
On a second thought:
1) Shouldn't we better check the type and fail if it's not an integer? The only accepted values/types would be integers, boolean, null and undefined.
2) `dlopen` can't be called with a zero flag, so I was treating the `0` as if the user requested passed `null`, allowing to fallback to default.

mscdex replied:
I suppose that's fine as well.

addaleax commented on src/node.cc:
```diff
@@ -2502,18 +2506,44 @@ struct node_module* get_linked_module(const char* name) {
+#ifdef __POSIX__
+  // If no flag is passed, fallback to default
+  if (args.Length() < 3 || args[2]->IsUndefined() ||
+      args[2]->IsNull() || args[2]->IntegerValue() == 0) {
```
`IntegerValue()` without arguments is deprecated, so it would be good if you could either use the overload that returns a `Maybe<int64_t>` and pass along the error if one occurs, or alternatively throw a type error if `args[2]->IsInt32()` fails and then use `args[2].As<Int32>()->Value()` to get the flags

ezequielgarcia replied:
how about something like:
```
  int flags = 0;
  if (args.Length() > 2) {
    if (!args[2]->IsInt32() && !args[2]->IsUndefined()) {
      return env->ThrowTypeError("dlopen flags must be undefined or integer.");
    }
    if (args[2]->IsInt32())
      flags = args[2]->Int32Value();
  }

  if (!flags) {
    // If no flag is passed, fallback to default
    is_dlopen_error = uv_dlopen(*filename, &lib);
  } else {
[the dlopen with flags]
```

jasnell commented on doc/api/process.md:
````diff
@@ -683,6 +683,22 @@ process.on('warning', (warning) => {
+process.dlopenFlags = os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_GLOBAL;
+var foo = require('foo');
+process.dlopenFlags = null;
+```
````
Not sure if a property is the right approach. I'd prefer a function (e.g. `setdlopenFlags()`/`getdlopenFlags()` that would include type checking on the value passed.

ezequielgarcia replied:
I'm fine either way. Whatever you guys decide it works better.

ezequielgarcia commented:
I don't how doable a `require.setdlopenFlags()` would be, but it might be the cleanest option. About the risks of different modules, I don't this is a big deal, if used properly. E.g.:
```
process.dlopenFlags = os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_GLOBAL;
var binding = require('bindings')('binding');
process.dlopenFlags = null;
```

sam-github commented:
> Question: should this be an API or a command line flag? What are the risks involved with different modules requiring different flags? Perhaps this is something that should be exposed off the module specific require instead (e.g. require.setdlopenFlags())? Just trying to think through the various ramifications.

^--- I had same questions as @jasnell from above. This seems it should be localized to a specific addon load, no?

ezequielgarcia commented:
@sam-github @jasnell Yes, it should be localized to one addon. However, it sounds that would require adding dlopenflags awareness to the `bindings` module. Any other idea?

gireeshpunathil commented:
Java virtual machines were plagued by namespace collision introduced through RTLD_GLOBAL - for example, an incoming symbol through dlopen could collide with an existing symbol in the already loaded images, and load order and other dependancy constraints come into play. To make things worse, JVM could be launched as a shared library from an existing process with already loaded libraries and symbols, and then the load order reverses, and the behavior too.

On the other hand, I recognize the use case, and the need to be able to expose loaded symbols outside of the library confinement. The current approach of platform default not only is restrictive, but also is platform dependant.

In my opinion, (i) the option to load addons with RTLD_GLOBAL should be localized on a per module basis to reduce side effects(ii) the functionality should be made platform neutral to reduce behavioral inconsistencies.

ezequielgarcia commented:
Hey guys,

First of all, thanks for the great feedback. I'm having lots of fun hacking node!

Just updated the branch, exploring the `require(..., opts)` route, as per the discussion in https://github.com/nodejs/node/issues/12848. Hopefully, I've addresses all the feedback received around API usage and coding style.

jasnell commented on lib/module.js:
```diff
@@ -447,6 +447,10 @@ Module._load = function(request, parent, isMain) {
   }
 
   var module = new Module(filename, parent);
+  if (opts !== undefined && opts.dlopenFlags) {
```
perhaps `if (opts !== null && typeof opts === 'object' && opts.dlopenFlags)` to make the check more precise.

as it is here, `require('foo', null) would throw an error

ezequielgarcia replied:
note that 'object' type is checked below. I was expecting `Module._load` to be called strictly with either 'undefined' or 'object'. Being internal, I thought I could expect that. let me know if that sounds ok.

jasnell replied:
Unfortunately that is not a safe assumption. `assert` should only be used when we know with absolute certainty that the value being tested comes from Node.js itself. In this case, the value comes from the user. Further, this test will fail if `opts === null` since `typeof null === 'object'` and `null !== undefined`

jasnell commented on lib/module.js:
```diff
@@ -447,6 +447,10 @@ Module._load = function(request, parent, isMain) {
 
   var module = new Module(filename, parent);
+  if (opts !== undefined && opts.dlopenFlags) {
+    assert(util.isNumber(opts.dlopenFlags), 'dlopenFlags must be a number');
```
This should not be an assert since this is checking a user provided value.
Instead, you'll want something like:

```
if (!util.isNumber(opts.dlopenFlags)) {
  const errors = require('internal/errors');
  throw new errors.TypeError('ERR_INVALID_OPT_VALUE',
                             'dlopenFlags', opts.dlopenFlags);
}
```

jasnell commented on lib/module.js:
```diff
@@ -447,6 +447,10 @@ Module._load = function(request, parent, isMain) {
   var module = new Module(filename, parent);
+  if (opts !== undefined && opts.dlopenFlags) {
+    assert(util.isNumber(opts.dlopenFlags), 'dlopenFlags must be a number');
+    module._dlopenFlags = opts.dlopenFlags;
```
Consider using a Symbol here instead of a new `_` prefixed property.

jasnell commented on lib/module.js:
```diff
@@ -520,10 +524,11 @@ Module.prototype.load = function(filename) {
   assert(path, 'missing path');
   assert(typeof path === 'string', 'path must be a string');
-  return Module._load(path, this, /* isMain */ false);
+  assert(opts === undefined || typeof opts === 'object', 'opts must be an object');
```
As above, this should not be an assert since it is testing user input. A TypeError should be used instead.

ezequielgarcia replied:
@jasnell Makes sense. Should the `assert` on the path argument be changed to exceptions too?

ezequielgarcia replied:
Also, I'm really wondering if checking the argument here makes any sense at all. The type check at `_load` might be just enough.

ezequielgarcia commented:
@jasnell @addaleax @mscdex Just updated the PR addressing James' feedback,

mscdex commented on doc/api/modules.md:
```diff
@@ -428,6 +428,29 @@ These are mostly for historic reasons.
+
+<!-- type=misc -->
+
+require() supports an optional parameter than can be used to pass
```
s/request()/\`require()\`/

mscdex commented on doc/api/modules.md:
```diff
@@ -428,6 +428,29 @@ These are mostly for historic reasons.
+<!-- type=misc -->
+
+require() supports an optional parameter than can be used to pass
+an object containing modifiers to the require operation.
```
s/require operation/module loader/ ?

mscdex commented on doc/api/modules.md:
```diff
@@ -428,6 +428,29 @@ These are mostly for historic reasons.
+an object containing modifiers to the require operation.
+
+Currently, the only property supported is `dlopenFlags`, which
+can be used to pass flags to the `dlopen` call that loads
```
This sentence sounds a bit odd to me. Perhaps reword/restructure the whole section to something like:

>Valid properties:
>
>* `dlopenFlags` {integer} dlopen(3) flags to be used when loading a Node.js addon. Example:
>
>    ```js
>    var dlopen = require('os').constants.dlopen;
>    var foo = require('./foo.node', { dlopenFlags: dlopen.RTLD_NOW | dlopen.RTLD_GLOBAL });
>    ```

bnoordhuis commented on doc/api/os.md:
```diff
@@ -1160,6 +1160,46 @@ The following error codes are specific to the Windows operating system:
+  </tr>
+  <tr>
+    <td><code>RTLD_DEEPBIND</code></td>
+    <td>Make a self-contained library use its own symbols in preference to symbols from previously loaded libraries.</td>
```
Long lines in this file, please wrap at 80 columns.

bnoordhuis commented on doc/api/modules.md:
````diff
@@ -428,6 +428,29 @@ These are mostly for historic reasons.
+
+```js
+var os = require('os');
+var foo = require('./foo', {dlopenFlags: os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_GLOBAL})
````
Long line, please wrap at 80 columns.

bnoordhuis commented on doc/api/os.md:
```diff
@@ -1160,6 +1160,46 @@ The following error codes are specific to the Windows operating system:
 
+### dlopen Constants
+
+The following signal constants are exported by `os.constants.dlopen`.
```
'are' is too definitive, they are only exported when available.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,18 +2491,51 @@ struct node_module* get_linked_module(const char* name) {
+    // libuv does not support passing flags to dlopen,
+    // so we nee to open code the uv_dlopen implementation.
+    dlerror();
+    lib.errmsg = NULL;
```
nullptr

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,18 +2491,51 @@ struct node_module* get_linked_module(const char* name) {
+      const char* errmsg = dlerror();
+
+      if (errmsg) {
+        lib.errmsg = strdup(errmsg);
```
This relies a bit too much on libuv implementation details to my taste.

bnoordhuis commented on test/parallel/test-binding-constants.js:
```diff
@@ -9,7 +9,7 @@ assert.deepStrictEqual(
 
 assert.deepStrictEqual(
-  Object.keys(constants.os).sort(), ['UV_UDP_REUSEADDR', 'errno', 'signals']
+  Object.keys(constants.os).sort(), ['UV_UDP_REUSEADDR', 'dlopen', 'errno', 'signals']
```
Long line.  Please run `make test` or `make lint`, that's what they are for.

ezequielgarcia commented:
@bnoordhuis How about this: https://github.com/ezequielgarcia/node-libao/blob/dlopen-hack/index.js#L7 ?

Not sure how problematic, or how nice, it is call `process.dlopen` directly in the module's index.js, but it does solve the problem and only needs an extra arg in `dlopen`. No require changes.

bnoordhuis commented:
@ezequielgarcia Something like that, although I don't understand why you pass in `module`.  I would have expected a two-arg overload of `process.dlopen()`.

ezequielgarcia commented:
@bnoordhuis You say an overload that uses the current module? since `process` is a native module, I thought it wasn't possible to overload a function from it. Maybe I was wrong.

bnoordhuis commented:
'Overload' in the sense that it (`process.dlopen`) would become a variadic function that takes one or two arguments.

ezequielgarcia commented:
So, currently `process.dlopen` expects a module and a filename. It `dlopen`s the filename, and uses the `module.exports` as placeholder for the functions exported by the filename.

We are going to be extending `process.dlopen` with a third argument, that expects ORed dlopen flags. It'd be `process.dlopen(module, filename, flags)`.

@bnoordhuis How would a different prototype work ? And how would we register the 
`dlopen`ed functions without a module?

bnoordhuis commented:
Oh, you're right.  Once upon a time it took a single argument and returned an object but you're right that it takes two now.  Objection withdrawn.

You can check the number of arguments with `args.Length()`.  The logic would be something like this:
```c++
int32_t mode = /* Default flags. */;
if (args.Length() >= 3 && !info[2]->Int32Value(env->context()).To(&mode))
  return;  // Exception pending.
```

ezequielgarcia commented:
@bnoordhuis OK, cool. I'll address your past comments and see if I can update the PR.
BTW, this PR is currently doing something like this:
```
 if (args.Length() > 2) {
    if (!args[2]->IsInt32() && !args[2]->IsUndefined()) {
      return env->ThrowTypeError("dlopen flags must be undefined or integer.");
    }
    if (args[2]->IsInt32())
      flags = args[2]->Int32Value();
}
```

ezequielgarcia commented:
@jasnell @addaleax @mscdex @bnoordhuis @gireeshpunathil @sam-github Updated the PR last night.

As discussed with @bnoordhuis, I've dropped all the changes to `require`, leaving only the dlopen constants and `process.dlopen` optional flag argument. This way, changes are not too invasive, while providing a solution that allows to set dlopen flags for a specific addon.

[Here](https://github.com/ezequielgarcia/node-libao/tree/dlopen-hack)'s an example on how this new interface can help a real life module. Without proper dlopen flags, libao is unable to load the ALSA plugin.

addaleax commented:
@ezequielgarcia This looks good from a glance; can you add documentation and, ideally, tests?

ezequielgarcia commented:
@addaleax How about a test such as:

```

addaleax commented:
@ezequielgarcia Something like that sounds okay, yes. Ideally, we’d have a test that actually can tell the difference between `RTLD_LOCAL` and `RTLD_GLOBAL`, but I get that that may be tricky.

mscdex commented on doc/api/os.md:
```diff
@@ -1160,6 +1160,52 @@ The following error codes are specific to the Windows operating system:
 
+### dlopen Constants
+
+If available on the operating system, the following signal constants
```
s/signal//

mscdex commented on doc/api/os.md:
```diff
@@ -1160,6 +1160,52 @@ The following error codes are specific to the Windows operating system:
+### dlopen Constants
+
+If available on the operating system, the following signal constants
+are exported by `os.constants.dlopen`. See [man 3 dlopen][] for detailed
```
s/by/in/

mscdex replied:
Also, the way man page links are handled is to use the format like 'dlopen(3)' and it will get automatically linked when the markdown is processed.

ezequielgarcia replied:
@mscdex Can you point me at any examples of this? I can't find any, but [man 3 foo][] usage.

mscdex replied:
Look at the `fs` documentation, that's the largest set of man page references.

mscdex commented:
Marking as semver-major due to the error text change.

bnoordhuis commented on src/node.cc:
```diff
@@ -2479,26 +2483,58 @@ struct node_module* get_linked_module(const char* name) {
+      return env->ThrowTypeError("dlopen flags must be undefined or integer.");
+    }
+    if (args[2]->IsInt32())
+      flags = args[2]->Int32Value();
```
Use the overload that takes a context (and handle errors.)

ezequielgarcia replied:
@bnoordhuis Would something like this (pattern taken from `node_crypto.cc`) work ?
```c++
  Maybe<int32_t> maybe_flags;
  if (args.Length() > 2) {
    CHECK(args[2]->IsInt32());
    maybe_flags = args[2]->Int32Value(env->context());
    CHECK(maybe_flags.IsJust());
  }
  // and later ....
  lib.handle = dlopen(*filename, maybe_flags.FromMaybe(RTLD_LAZY));
```

BTW, Isn't the first `CHECK(args[2]->IsInt32());` redundant since we are using a Maybe type?

bnoordhuis replied:
Yes and yes (yes, works and yes, redundant).  An alternative is this:
```c++
int32_t flags = RTLD_LAZY;
if (args.Length() > 2 && !args[2]->Int32Value(env->context()).To(&flags)) {
  // handle error
}
```

addaleax replied:
> BTW, Isn't the first `CHECK(args[2]->IsInt32());` redundant since we are using a Maybe type?

Kind of. The `Maybe` is used to indicate errors that happen during a possible conversion to the JS `number` type (for example because calling `[Symbol.toPrimitive]` on `args[2]` throws an exception), not to tell you whether `args[2]` is an Int32 or not.

Basically, there are two ways to do this:
1. If you use `CHECK(args[2]->IsInt32());`, that means you know that `args[2]` is an Int32, and you can use `args[2].As<Int32>()->Value()` to get its `int32_t` interpretation.
2. If you use `args[2]->Int32Value(env->context())`, that means that `args[2]` can be any kind of JS value, which will converted to a `number` based on JavaScript’s conversion rules. Because that operation can run userland code, it can fail, and in that case you get an empty `Maybe<int32_t>`.

bnoordhuis commented on src/node.cc:
```diff
@@ -2479,26 +2483,58 @@ struct node_module* get_linked_module(const char* name) {
+      flags = args[2]->Int32Value();
+  }
+
+  if (!flags) {
```
Style nit: `flags == 0`.  Rule of thumb: `!` only for bools.

bnoordhuis commented on src/node.cc:
```diff
@@ -2479,26 +2483,58 @@ struct node_module* get_linked_module(const char* name) {
+      is_dlopen_error = true;
+      dlerrmsg = dlerror();
+    }
+  }
```
I'd just use dlopen() unconditionally, one less code path to reason about.

bnoordhuis commented on test/addons/dlopen-hello-world/binding.cc:
```diff
@@ -0,0 +1,14 @@
+
+void Method(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+  v8::HandleScope scope(isolate);
```
No HandleScope necessary.  I thought I cleaned those up a while ago but apparently I missed hello-world/binding.cc...

ezequielgarcia replied:
OK - I can send a PR removing those while here.

bnoordhuis commented on test/addons/dlopen-hello-world/test.js:
```diff
@@ -0,0 +1,8 @@
+const assert = require('assert');
+const os = require('os');
+process.dlopen(module, require.resolve(`./build/${common.buildType}/binding`),
+               os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_LOCAL)
```
I guess this works by accident on Windows?  `undefined | undefined` is `NaN | NaN` is `0 | 0` is `0`.

In that respect this test doesn't really test anything, it will work whether the dlopen flags are respected or not.

Alternative: two add-ons, first one loaded with RTLD_GLOBAL, second one linking to a symbol defined in the first one (and skip the test on Windows.)

ezequielgarcia replied:
Wasn't an accident, I was kind of relying on that behavior to avoid skipping the test. Yeah, the alternative sounds like a good idea.

ezequielgarcia commented:
@mscdex @bnoordhuis @addaleax Addressed your feedback. Also, noticed a problem in `test-constants.js` and fixed it.

Let me know how the new dlopen-ping-pong test looks. The test passes only if proper support for global symbol loading is available (i.e. RTLD_GLOBAL dlopen flag).

addaleax commented on doc/api/process.md:
```diff
@@ -637,6 +637,28 @@ process's [`ChildProcess.disconnect()`][].
 
+## process.dlopen(module, filename[, flags])
+<!-- YAML
+added: REPLACEME
```
Can you do something like
```yaml
added: v0.1.8
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/12794
    description: Added support for the `flags` argument.
```
?

ezequielgarcia replied:
@addaleax `process.dlopen` was added in v2.0.0. It was called `node.dlopen` before I think. What should we claim?

addaleax replied:
```
$ nvm use 0.5
Now using node v0.5.10
$ node -e 'console.log(process.dlopen)'
[Function]
```

My nvm doesn’t go back further than 0.5, but it was definitely there before. The version we should insert here is the first version in which it was available as `process.dlopen`, which I think is 0.1.8, but at those version ranges it probably doesn’t matter much anyway.

ezequielgarcia replied:
Gah, bad git-magic. I believe it was moved from `node.dlopen` to `process.dlopen` in ad0a4cefb8b4bf469bb1caa6e671a1009e40a81f, which made it to v0.1.16. Also v0.1.16 commit 726865af7bbafe58435986f4a193ff11c84e4bfe confirms it.

addaleax commented on test/addons/dlopen-ping-pong/test.js:
```diff
@@ -0,0 +1,18 @@
+var bindingPath = require.resolve(`./build/${common.buildType}/binding`)
+process.dlopen(module, bindingPath,
+               os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_GLOBAL)
+module.exports.load(path.dirname(bindingPath) + '/ping.so')
```
I don’t think this passes the linter? We require semicolons ;)

ezequielgarcia replied:
oh, boy! i wonder what happened there...

addaleax commented on test/parallel/test-constants.js:
```diff
@@ -14,7 +14,7 @@ assert.ok(binding.crypto);
 ['os', 'fs', 'crypto'].forEach((l) => {
   Object.keys(binding[l]).forEach((k) => {
-    if (typeof binding[l][k] === 'object') { // errno and signals
+    if (k === 'errno' || k === 'signals') {
```
Not sure, why this change?

ezequielgarcia replied:
Because this test fails without an explicit check for `errno` and `signals`. `process.binding('constants')` exports constants from `node_constants.cc:DefineConstants()`, while `require('constants')` exports constants from the seemingly deprecated lib/constants.js.

I haven't touched lib/constants.js because it said it was deprecated, and so now the exports don't match. That make sense?

bnoordhuis replied:
I'd update lib/constants.js.  It'd be odd for them to be out of sync, IMO.

refack commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8786/

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,26 +2491,48 @@ struct node_module* get_linked_module(const char* name) {
+  if (!lib.handle) {
+    is_dlopen_error = true;
+    dlerrmsg = dlerror();
+  }
```
This still mixes libuv code with direct calls, which I don't think is a good idea.  I'd use libuv on Windows and plain dlopen/dlerror on UNIX.  Maybe split it off into separate functions to avoid any confusion.

If it makes it easier, you can remove the "Module did not self-register" uv_dlclose() call because I don't think that one is sound.  Code from the shared object can have run so it's not safe to unload it again.

ezequielgarcia replied:
OK, a split makes sense.

Regarding "Module did not self-register" uv_dlclose() , I'm not sure what you mean by "unload it again". The module is only unloaded on that path, AFAICS.

bnoordhuis replied:
dlclose-ing a shared object is safe only when you know it never executed any code (but we don't know that, there may be static constructors) or when you are absolutely sure it cleaned up after itself (which we don't know either.)

ezequielgarcia replied:
@bnoordhuis Right, but in that case, shouldn't we remove all the `dlclose` calls? Sorry, but I can't see how the one at "Module did not self-register" is special.

ezequielgarcia replied:
Thinking some more about this, that doesn't make much sense to me. Resources allocated by static constructors should be released by static deconstructors. I'd say it's safe to call `dlclose` as long as you don't use exported (i.e. `dlsym`ed) symbols. The current code looks good in that regard.

On the other side, I'd be happy to tackle this if it's an issue, but please let's not scope creep this PR, which is already enough controversial on its own ;-)

bnoordhuis replied:
That's why I said "if it makes it easier."  If it doesn't make it easier, by all means skip it.

> Resources allocated by static constructors should be released by static deconstructors.

Should, yes, but you can't assume that actually happens when loading arbitrary code.

bnoordhuis commented on test/addons/dlopen-ping-pong/binding.cc:
```diff
@@ -0,0 +1,43 @@
+
+typedef const char* (*ping_t)(void);
+
+static ping_t pingFunc;
```
Style nit: `ping_func`

bnoordhuis commented on test/addons/dlopen-ping-pong/ping.c:
```diff
@@ -0,0 +1,6 @@
+const char* dlopen_ping(void) {
+  return dlopen_pong();
+}
+
```
Style nit: no blank line here.

bnoordhuis commented on test/addons/dlopen-ping-pong/binding.gyp:
```diff
@@ -0,0 +1,15 @@
+    {
+      'target_name': 'ping',
+      'type': 'shared_library',
+      'defines': [ 'V8_DEPRECATION_WARNINGS=1' ],
```
Forgot to mention, this define is not strictly necessary.  It doesn't hurt either, though.

ezequielgarcia commented:
@bnoordhuis how about that for the implementation?

btw, i'm thinking i can add some tests that fail on the dlopen (loading a file that is not there, or that is not a shared library, or that is not a node addon).

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,46 +2491,100 @@ struct node_module* get_linked_module(const char* name) {
+    return handle;
+
+  const char* errmsg = dlerror();
+  *dlerrmsg = strdup((errmsg != nullptr) ? errmsg : "no error");
```
I'd make this a `std::string*` and then simply do `if (errmsg != nullptr) *dlerrmsg = errmsg;`  The less manual memory allocation, the better.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,46 +2491,100 @@ struct node_module* get_linked_module(const char* name) {
+
+#else  // _WIN32
+
+static void* node_dlhandle(void *handle) {
```
Style (here and below): star leans left.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,46 +2491,100 @@ struct node_module* get_linked_module(const char* name) {
+#else  // _WIN32
+
+static void* node_dlhandle(void *handle) {
+  uv_lib_t* lib = reinterpret_cast<uv_lib_t*>(handle);
```
You can just `static_cast<...>` here and below.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,46 +2491,100 @@ struct node_module* get_linked_module(const char* name) {
+  uv_dlclose(lib);
+  delete lib;
+  return nullptr;
+}
```
Can I suggest turning these functions into a RAII class?  That way the uv_lib_t can simply be a data member in the Windows implementation and no juggling of new/delete calls is necessary.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,46 +2491,100 @@ struct node_module* get_linked_module(const char* name) {
     return;
   }
 
+  int32_t flags = RTLD_LAZY;
```
Does this actually compile on Windows?

refack replied:
Nope:
```console
src\node.cc(2558): error C2065: 'RTLD_LAZY': undeclared identifier [p:\code\node\node.vcxproj]
```

ezequielgarcia commented:
@bnoordhuis How about it now? I did a class to abstract the shared object load. Now there are no explicit allocations ;-)

jasnell commented on doc/api/process.md:
```diff
@@ -637,6 +637,32 @@ process's [`ChildProcess.disconnect()`][].
+
+The `process.dlopen()` method allows to load dynamically shared
+objects. It is primarily used by `require()` to load
+C++ Addons, and should not be used directly expect in special
```
s/expect/except

jasnell commented on doc/api/process.md:
````diff
@@ -637,6 +637,32 @@ process's [`ChildProcess.disconnect()`][].
+
+```js
+process.dlopen(module, require.resolve('binding'));
+```
````
A bit of text explaining the use of `require.resolve()` here may be helpful. As it is, this example is a bit terse.
Also, some text about the possible tradeoffs are with bypassing `require()` in this way would be helpful.

jasnell replied:
Also, an example that shows use of the `flags` would be helpful

ezequielgarcia replied:
@jasnell how about it now?

jasnell commented on test/addons/dlopen-ping-pong/test.js:
```diff
@@ -0,0 +1,18 @@
+               os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_GLOBAL);
+module.exports.load(path.dirname(bindingPath) + '/ping.so');
+assert.strictEqual(module.exports.ping(), 'pong');
+console.log('binding.ping() =', module.exports.ping());
```
It would be helpful for the test to also verify what happens when the module is `required()'d after the `process.dlopen()` call.

ezequielgarcia replied:
@jasnell You sure that would belong on this test? The result will be an exception with error `Error: Module did not self-register.`.
If needed, I'd say it goes in its own test, but to be honest I'm not entirely sure what's the value of it.

jasnell replied:
The value is purely completeness of the test. It is entirely feasible that a user (or some set of dependencies) to do this kind of thing and it would be helpful to have verification of the expected behavior.

bnoordhuis replied:
Can you drop the console.log?  Tests should be silent (or as silent as possible anyway.)

ezequielgarcia commented:
@jasnell Just updated the doc again, rephrasing it. Hope it's clearer now!

jasnell commented on doc/api/process.md:
```diff
@@ -637,6 +637,48 @@ process's [`ChildProcess.disconnect()`][].
+
+One important drawback when calling `process.dlopen` is that you need
+to pass a `module` instance. Functions exported by the Addon,
+are acessible via `module.exports`.
```
please avoid the use of informal pronouns like `you` in the docs. This could be reworded as:

```
*Note*: An important drawback when calling `process.dlopen()` is that the
`module` instance must be passed. Functions exported by the addon will
be accessible via `module.exports`.
```

jasnell commented on doc/api/process.md:
```diff
@@ -637,6 +637,48 @@ process's [`ChildProcess.disconnect()`][].
+The example below shows how to load a C++ Addon, named as `binding`,
+that exports a `foo` function. All the symbols will be loaded before
+the call returns, by passing the `RTLD_NOW` constant. In this example
+the constant is assumed to be available on your platform.
```
```
In this example, the constant is assume to be available.
```

ezequielgarcia commented:
@jasnell ok, fixed.

ezequielgarcia commented on doc/api/process.md:
```diff
@@ -637,6 +637,48 @@ process's [`ChildProcess.disconnect()`][].
+The `process.dlopen()` method allows to dynamically load shared
+objects. It is primarily used by `require()` to load
+C++ Addons, and should not be used directly, except in special
+cases. In other words, unless you have a good reason, you should
```
oh, there are more *you* uses. I have to clean this up.

ezequielgarcia commented:
@bnoordhuis @jasnell Updated docs again, no more `you` uses. Let me know how it looks.

bnoordhuis commented on doc/api/process.md:
```diff
@@ -637,6 +637,48 @@ process's [`ChildProcess.disconnect()`][].
+
+If there are specific reasons to use `process.dlopen()` (for instance,
+to specify dlopen flags), it's often useful to use [`require.resolve()`][]
+to lookup the module's path.
```
I think this should be 'to look up'?

ezequielgarcia replied:
Yep.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,36 +2491,113 @@ struct node_module* get_linked_module(const char* name) {
+  void Close();
+  const char* GetError();
+
+  operator void*() const {;
```
Can you drop the semicolon?

ezequielgarcia replied:
Oops.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,36 +2491,113 @@ struct node_module* get_linked_module(const char* name) {
+    return true;
+
+  const char* errmsg = dlerror();
+  if (errmsg != nullptr)
```
nullptr check shouldn't be necessary, just `errmsg_ = dlerror();` should suffice.

ezequielgarcia replied:
Done.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,36 +2491,113 @@ struct node_module* get_linked_module(const char* name) {
+}
+
+void DLib::Close() {
+  if (this->handle_)
```
Can you check `this->handle_ != nullptr`?

ezequielgarcia replied:
Done.

bnoordhuis commented on src/node.cc:
```diff
@@ -2553,18 +2634,18 @@ static void DLOpen(const FunctionCallbackInfo<Value>& args) {
   }
 
-  mp->nm_dso_handle = lib.handle;
+  mp->nm_dso_handle = static_cast<void*>(dlib);
```
This looks confusing.  I'd do away with the cast operator overloading and replace it with a `dlib.handle()` call.

Since it's a small, one-off class, making it a simple struct and accesssing data members directly is acceptable too, IMO.  That will let you shrink it some more.

ezequielgarcia replied:
Done. I hope it looks acceptable now.

bnoordhuis commented on test/addons/dlopen-ping-pong/binding.cc:
```diff
@@ -0,0 +1,43 @@
+
+NODE_MODULE(binding, init)
+
+}  // anonymous namespace
```
I suspect that you need to wrap this whole file in a `#ifndef _WIN32 ... #endif` guard.  The CI buildbots build all add-ons irrespective of platform, but this file won't build on Windows.

ezequielgarcia replied:
Done.

bnoordhuis commented on src/node.cc:
```diff
@@ -2487,36 +2491,113 @@ struct node_module* get_linked_module(const char* name) {
+
+void DLib::Close() {
+  if (this->handle_)
+    dlclose(this->handle_);
```
And set `this->handle_ = nullptr;`?

Side note: `this->` isn't necessary and we don't usually use that style in core.  The fact that the field's name ends in an underscore signifies that it's a data member, not a local.

Trott commented:
@ezequielgarcia Still working on this?

@bnoordhuis @jasnell Consider maybe pushing your nit fixes to their branch as `squash` commits?

ezequielgarcia commented:
@Trott Haven't forgot about this one, but got a little distracted by other stuff. It's on my queue though.

ezequielgarcia commented:
@bnoordhuis @jasnell @Trott Sorry for the delay!

Here's a new round! Could it be the final one? 🤔 
Biggest change is rework of the DLib type: using a struct abstraction instead of a class. Tests passed, Linux only. Also, fixed all nits pointed out by @bnoordhuis.

Trott commented:
@bnoordhuis @jasnell Your requests for changes are the blockers on this right now. Any chance you can take a look?

ezequielgarcia commented:
@bnoordhuis @jasnell @Trott @addaleax @mscdex @refack How about we spin a new ci build?

jasnell commented:
Will do so in the morning if someone doesn't beat me to it

bnoordhuis commented on src/node.cc:
```diff
@@ -2497,36 +2501,95 @@ struct node_module* get_linked_module(const char* name) {
+  std::string errmsg_;
+  int flags_;
+
+  void* Handle(struct DLib *lib) {
```
That's an... unusual idiom.  Make it static (but why?) or drop the parameter and do `return handle_;` directly.

ezequielgarcia replied:
Hm, to be honest, I wasn't familiar at all with C++ structs and I assumed they worked like C structs. Shame on me.

bnoordhuis commented on src/node.cc:
```diff
@@ -2497,36 +2501,95 @@ struct node_module* get_linked_module(const char* name) {
+
+#ifdef __POSIX__
+
+  #define DLOPEN_DEFAULT_FLAGS RTLD_LAZY;
```
Don't indent preprocessor directives (not portable) and drop the semicolon.

Namespace it by calling it `NODE_DLOPEN_DEFAULT_FLAGS` or better, turn it into `static const int kDefaultFlags = RTLD_LAZY;`.

ezequielgarcia replied:
Oh, of course.

ezequielgarcia commented:
@bnoordhuis @jasnell @Trott @addaleax @mscdex @refack Comments from @bnoordhuis addressed and conflict solved.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9868/

jasnell commented:
Looks like this is failing on OSX and has lint issues to address

ezequielgarcia commented:
@jasnell I've addressed the linter error. Regarding OSX, here's the error:

```
Undefined symbols for architecture x86_64:
  "_dlopen_pong", referenced from:
      _dlopen_ping in ping.o
```
I think some ifdef-ery should do it:
```
#ifndef _WIN32

#ifdef __APPLE__
const char* _dlopen_pong(void);

const char* dlopen_ping(void) {
  return _dlopen_pong();
}
#else
const char* dlopen_pong(void);

const char* dlopen_ping(void) {
  return dlopen_pong();
}
#endif  // __APPLE__

#endif  // _WIN32
```

ezequielgarcia commented:
@jasnell I think I got it. Any chance we fire another ci? 🙏

jasnell commented:
Here we go: https://ci.nodejs.org/job/node-test-pull-request/9872/

jasnell commented:
btw... I really appreciate you sticking with this @ezequielgarcia ... I know it has been a long road, but the effort is totally appreciated.

ezequielgarcia commented:
The test is still not working correctly in OSX. The problem is not with the code, but it's the test itself, due to OSX putting an underscore prefix and me not knowing how to work with it.
Maybe we should just limit the test to Linux and friends? Thoughts @jasnell @Trott @bnoordhuis @addaleax et al ?

jasnell commented:
this would have to pass on OSX regardless. I'll take a look later on today and see if I can figure out what is happening.

ezequielgarcia commented:
@jasnell that would be much appreciated. I don't have access to OSX so it's hard for me to do it.

ezequielgarcia commented:
@jasnell I've fixed the OSX linker flags. Can you start another build?

jasnell commented:
I will but today is a holiday in the US so it may not be until tomorrow (pretty certain my wife would kill me if I dared touch the laptop today)

refack commented:
CI: https://ci.nodejs.org/job/node-test-commit/12189/

refack commented:
Some failures look related:
* https://ci.nodejs.org/job/node-compile-windows/11555/label=win-vs2015/console
doesn't seem to link only on win-vs2015 (server2012r2x64)
```
     Creating library c:\workspace\node-compile-windows\label\win-vs2015\Release\node.lib and object c:\workspace\node-compile-windows\label\win-vs2015\Release\node.exp
  Generating code
  Finished generating code
nghttp2.lib(nghttp2_stream.obj) : warning LNK4099: PDB 'nghttp2.pdb' was not found with 'nghttp2.lib(nghttp2_stream.obj)' or at 'c:\workspace\node-compile-windows\label\win-vs2015\Release\nghttp2.pdb'; linking object as if no debug info [c:\workspace\node-compile-windows\label\win-vs2015\node.vcxproj]
nghttp2.lib(nghttp2_session.obj) : warning LNK4099: PDB 'nghttp2.pdb' was not found with 'nghttp2.lib(nghttp2_session.obj)' or at 'c:\workspace\node-compile-windows\label\win-vs2015\Release\nghttp2.pdb'; linking object as if no debug info [c:\workspace\node-compile-windows\label\win-vs2015\node.vcxproj]
nghttp2.lib(nghttp2_callbacks.obj) : warning LNK4099: PDB 'nghttp2.pdb' was not found with 'nghttp2.lib(nghttp2_callbacks.obj)' or at 'c:\workspace\node-compile-windows\label\win-vs2015\Release\nghttp2.pdb'; linking object as if no debug info [c:\workspace\node-compile-windows\label\win-vs2015\node.vcxproj]
...
```

* https://ci.nodejs.org/job/node-test-commit-osx/12186/nodes=osx1010/console
maybe changed abort behaviour (maybe just a macOS flake)
```
not ok 77 async-hooks/test-callback-error
  ---
  duration_ms: 15.570
  severity: fail
  stack: |-
    start case 1
    end case 1: 149.323ms
    start case 2
    end case 2: 87.329ms
    start case 3
    end case 3: 10.827ms
    Error: test_callback_abort
        at ActivityCollector.initHooks.oninit.common.mustCall (/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/test/async-hooks/test-callback-error.js:36:45)
        at ActivityCollector.oninit (/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/test/common/index.js:509:15)
        at ActivityCollector._init (/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/test/async-hooks/init-hooks.js:182:10)
        at emitInitNative (async_hooks.js:446:43)
        at Object.emitInitScript [as emitInit] (async_hooks.js:349:3)
        at Object.<anonymous> (/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/test/async-hooks/test-callback-error.js:38:17)
        at Module._compile (module.js:549:30)
        at Object.Module._extensions..js (module.js:560:10)
        at Module.load (module.js:483:32)
        at tryModuleLoad (module.js:446:12)
     1: node::Abort() [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
     2: node::Chdir(v8::FunctionCallbackInfo<v8::Value> const&) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
     3: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
     4: v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
     5: v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
     6: 0x3845705046fd
    
  ...
```

* https://ci.nodejs.org/job/node-test-commit-aix/8330/nodes=aix61-ppc64/console
AIX doesn't seem to link correctly:
```
  g++ -pthread -Wl,-bbigtoc -maix64 -Wl,-bE:/home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/node.exp -Wl,-blibpath:/usr/lib:/lib:/opt/freeware/lib/pthread/ppc64 -pthread  -o /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/node /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/node/src/node_main.o /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/libnode_base.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/nghttp2/libnghttp2.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/v8/src/libv8_base.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/v8/src/libv8_libbase.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/v8/src/libv8_libsampler.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/tools/icu/libicui18n.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/tools/icu/libicuucx.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/tools/icu/libicudata.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/tools/icu/libicustubdata.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/v8/src/libv8_snapshot.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/v8/src/libv8_libplatform.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/openssl/libopenssl.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/zlib/libzlib.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/http_parser/libhttp_parser.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/cares/libcares.a /home/iojs/build/workspace/node-test-commit-aix/nodes/aix61-ppc64/out/Release/obj.target/deps/uv/libuv.a -lm -lperfstat
ld: 0711-783 WARNING: TOC overflow. TOC size: 88296	Maximum size: 65536
	Extra instructions are being generated for each reference to a TOC
	symbol if the symbol is in the TOC overflow area.
ld: 0711-319 WARNING: Exported symbol not defined: v8::internal::interpreter::SetupInterpreter::ReuseExistingHandler(unsigned char**, v8::internal::interpreter::Bytecode, v8::internal::interpreter::OperandScale)
ld: 0711-319 WARNING: Exported symbol not defined: v8::internal::interpreter::SetupInterpreter::InstallBytecodeHandler(v8::internal::Isolate*, unsigned char**, v8::internal::interpreter::Bytecode, v8::internal::interpreter::OperandScale)
ld: 0711-319 WARNING: Exported symbol not defined: v8::internal::interpreter::SetupInterpreter::InstallBytecodeHandlers(v8::internal::interpreter::Interpreter*)
...
```

refack commented:
Resume CI: https://ci.nodejs.org/job/node-test-commit/12191/

refack commented:
https://ci.nodejs.org/job/node-test-commit/12198/

ezequielgarcia commented:
Looks like those failures aren't related strictly to this change set. Is there any other feedback?

gireeshpunathil commented:
I don't think the AIX warnings are related to this, or even new. I have seen the `TOC overflow` related warnings often, that has todo with the linker's internal structures. Let me go through some of the contemporary builds to see if these were always there.

refack commented:
@gireeshpunathil @nodejs/platform-aix The issue on AIX is actually:
```make
gmake[1]: Entering directory `/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong/build'
  g++ '-DNODE_GYP_MODULE_NAME=binding' '-DUSING_UV_SHARED=1' '-DUSING_V8_SHARED=1' '-DV8_DEPRECATION_WARNINGS=1' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/include/node -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/src -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/deps/uv/include -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/deps/v8/include  -pthread -Wall -Wextra -Wno-unused-parameter -maix64 -O3 -fno-omit-frame-pointer -fno-rtti -fno-exceptions -std=gnu++0x -MMD -MF ./Release/.deps/Release/obj.target/binding/binding.o.d.raw   -c -o Release/obj.target/binding/binding.o ../binding.cc
gyp info spawn gmake
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  g++ -shared -Wl,-bimport:/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/out/Release/node.exp -pthread -Wl,-bbigtoc -maix64  -o Release/obj.target/binding.node Release/obj.target/binding/binding.o
  rm -rf "Release/binding.node" && cp -pPRf "Release/obj.target/binding.node" "Release/binding.node"
  gcc '-DNODE_GYP_MODULE_NAME=ping' '-DUSING_UV_SHARED=1' '-DUSING_V8_SHARED=1' '-DV8_DEPRECATION_WARNINGS=1' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/include/node -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/src -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/deps/uv/include -I/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/deps/v8/include  -pthread -Wall -Wextra -Wno-unused-parameter -maix64 -O3 -fno-omit-frame-pointer  -MMD -MF ./Release/.deps/Release/obj.target/ping/ping.o.d.raw   -c -o Release/obj.target/ping/ping.o ../ping.c
  g++ -shared -Wl,-bimport:/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/out/Release/node.exp -pthread -Wl,-bbigtoc -maix64  -o Release/obj.target/ping.so ./Release/obj.target/ping/ping.o
ld: 0711-317 ERROR: Undefined symbol: .dlopen_pong
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
gmake[1]: *** [Release/obj.target/ping.so] Error 1
gmake[1]: Leaving directory `/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong/build'
gyp ERR! build error
gyp ERR! stack Error: `gmake` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/deps/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:125:13)
gyp ERR! stack     at ChildProcess.emit (events.js:213:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:210:12)
gyp ERR! System AIX 6.1
gyp ERR! command "/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/out/Release/node" "/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/deps/npm/node_modules/node-gyp/bin/node-gyp" "--loglevel=info" "rebuild" "--python=/usr/bin/python" "--directory=/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong/" "--nodedir=/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp"
gyp ERR! cwd /home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong
gyp ERR! node -v v9.0.0-pre
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
gmake: *** [test/addons/.buildstamp] Error 1
```

We patched the build error with `'-Wl,-berok'` but that only turned it into a runtime error:
```cmd
bash-4.3# python tools/test.py --mode=release -J addons
=== release test ===
Path: addons/dlopen-ping-pong/test
Command: out/Release/node /home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong/test.js
--- CRASHED (Signal: 4) ---
[00:28|% 100|+  41|-   1]: Done
bash-4.3# out/Release/node /home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong/test.js
Illegal instruction (core dumped)
```

TimothyGu commented on lib/constants.js:
```diff
@@ -27,6 +27,7 @@
 // Deprecation Code: DEP0008
 const constants = process.binding('constants');
 Object.assign(exports,
+              constants.os.dlopen,
```
I don't think we should add things to something that's already deprecated.

ezequielgarcia replied:
This has already been discussed, on this same PR-thread. I'll let you do the digging. (BTW: this is where mailing lists outpass github PRs: it's so much harder to dig discussion history).

TimothyGu replied:
Okay, in that case SGTM. Sorry about not reading through the entire thread – though in my defense it is quite long.

TimothyGu commented on src/node_constants.cc:
```diff
@@ -1279,9 +1309,11 @@ void DefineConstants(v8::Isolate* isolate, Local<Object> target) {
   os_constants->Set(OneByteString(isolate, "errno"), err_constants);
   os_constants->Set(OneByteString(isolate, "signals"), sig_constants);
+  os_constants->Set(OneByteString(isolate, "dlopen"), dlopen_constants);
   target->Set(OneByteString(isolate, "os"), os_constants);
```
Alphabetical order

ezequielgarcia replied:
OK.

TimothyGu commented on test/addons/dlopen-ping-pong/binding.cc:
```diff
@@ -0,0 +1,48 @@
+using v8::String;
+using v8::Value;
+
+typedef const char* (*ping_t)(void);
```
`_t` suffix is technically reserved.

ezequielgarcia replied:
It's just a test. Does it matter?

TimothyGu replied:
It doesn't, that's why I said "technically" 😛

Anyways it's up to you if you want to change it.

addaleax replied:
@TimothyGu Just curious, wouldn’t that just be for C anyway? C++ has namespacing and this is even inside an anonymous one…

ezequielgarcia commented:
Just pushed a new branch fixing @TimothyGu nitpicks.

gireeshpunathil commented:
linker warnings 0711-783 was there for quite sometime and 0711-319 was introduced recently, but neither of them are new through this PR.

proof:
https://ci.nodejs.org/job/node-test-commit-aix/nodes=aix61-ppc64/8331/consoleFull

```c++
ld: 0711-783 WARNING: TOC overflow. TOC size: 67960	Maximum size: 65536
	Extra instructions are being generated for each reference to a TOC
	symbol if the symbol is in the TOC overflow area.
ld: 0711-783 WARNING: TOC overflow. TOC size: 68400	Maximum size: 65536
	Extra instructions are being generated for each reference to a TOC
	symbol if the symbol is in the TOC overflow area.
```
```c++
ld: 0711-319 WARNING: Exported symbol not defined: v8::internal::interpreter::SetupInterpreter::ReuseExistingHandler(unsigned char**, v8::internal::interpreter::Bytecode, v8::internal::interpreter::OperandScale)
```

The AIX errors seem to be root caused by some infra issues, predating this PR: I am seeing ~100 instances of this reasons for test failures.

`Error: EACCES: permission denied, unlink '/home/iojs/node-tmp/tmp.0/addon'`

Looks like a stale file from some older runs, which needs to be removed.

mhdawson commented:
The 2 TOC warnings are expected and normal.  They are not a problem.

ld: 0711-319 WARNING: Exported symbol not defined: v8::internal::interpreter::SetupInterpreter::ReuseExistingHandler(unsigned char**, v8::internal::interpreter::Bytecode, v8::internal::interpreter::OperandScale)

we probably need to investigate but if its not related to this PR as @gireeshpunathil mentioned, we should open a separate issue to investigate.

I'll take a look at '/home/iojs/node-tmp/tmp.0/addon'

mhdawson commented:
ok so for some reason on `test-osuosl-aix61-ppc64_be-1` /home/iojs/node-tmp/tmp.0/addon is owned by root which seems wrong.   on _be-2 is seems to be owned by iojs which makes sense

I'll remove the file and do some cleanup as it seems like there are stray jobs on the machine as well.

mhdawson commented:
CI run after clean, on this PR/AIX https://ci.nodejs.org/job/node-test-commit-aix/nodes=aix61-ppc64/8352/console

mhdawson commented:
Looks like that run was successful, Full CI run to be sure: https://ci.nodejs.org/job/node-test-pull-request/9964/

mhdawson commented:
OSX failure in latest CI run is not related to this PR (it was pre-existing).

refack commented:
@nodejs/platform-aix the tl;dr of this issue with AIX:

Original build error:
```
  g++ -shared -Wl,-bimport:/home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/out/Release/node.exp -pthread -Wl,-bbigtoc -maix64  -o Release/obj.target/ping.so ./Release/obj.target/ping/ping.o
ld: 0711-317 ERROR: Undefined symbol: .dlopen_pong
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
```
Circumvented with:
https://github.com/nodejs/node/blob/69e911e1455db214f9101ce966ae14e4baf1ea5d/test/addons/dlopen-ping-pong/binding.gyp#L13-L17

But this causes:
```
bash-4.3# out/Release/node /home/iojs/build/workspace/node-test-commit-aix/nodes/tmp/test/addons/dlopen-ping-pong/test.js
Illegal instruction (core dumped)
```

So test currently excludes AIX:
https://github.com/nodejs/node/blob/69e911e1455db214f9101ce966ae14e4baf1ea5d/test/addons/dlopen-ping-pong/test.js#L9-L12

---

P.S. the `/home/iojs/node-tmp/tmp.0/addon` addon issue is my fault, I was helping @ezequielgarcia by running tests on the machine.

richardlau commented:
re. AIX: 
```
ld: 0711-317 ERROR: Undefined symbol: .dlopen_pong
```

AIUI the AIX linker does not export all symbols by default, so the symbol needs to be explicitly exported from the new test library (i.e. don't ignore with _erok_, this is a real error). There are more than one way of doing this:
https://www.ibm.com/developerworks/aix/library/au-aix-symbol-visibility/index.html
For Node.js itself, we generate and use an export (_node.exp_) file:
https://github.com/nodejs/node/blob/8485a7c0b71fd1dd36947a6dd02b75387805d9b7/node.gyp#L728-L790

richardlau commented:
(Please note when looking at the link I posted that for Node.js we use the GNU C/C++ compiler (and not XL C/C++) but still use the standard AIX linker).

ezequielgarcia commented:
@richardlau Thanks for looking into it! I think you got it wrong. The test is purposely *not* linking, so the dlopen_pong symbol is *meant to be unresolved*.

The idea behind this test is that Node.js will load binding.node using `RTLD_GLOBAL` so that `dlopen_pong` is globally exported for later loaded libraries. Then, binding.node itself will load ping.so. Given `RTLD_GLOBAL` was usage, ping.so (loaded by binding.node) is able to access binding.node's symbol `dlopen_pong`. Makes sense or did I make a mess of it?

So, if the above made sense, then the conclusion is that at link time, ping.so should not resolve `dlopen_pong`, the symbol must remain unresolved. Please note that we are passing some flags in OSX to make this work. In Linux, it works by default.

I guess the test is a little tricky, but I haven't managed to find a simpler test for the dlopen `RTLD_GLOBAL` flag.

ezequielgarcia commented:
Quoting the explanation in the first comment, the test is basically exercising what some libraries that use a plugin architecture do:

> In particular, it's common to require a node addon to be loaded using dlopen(flag=RTLD_GLOBAL). For instance, this is required to implement a libao binding, due to libao's plugins being shared objects that need symbols exposed in a common shared object (see here).

refack commented:
@ezequielgarcia I think I saw something in the guide @richardlau pointed to:
```
/* generate position-independent code suitable for use in shared libraries. */
$ xlC -c func.C invoke.C -qpic
 
/* generate shared library, exportlist is used to control symbol visibility */
$ xlC -G -o libtest.so func.o invoke.o -bE:exportlist
 
$ xlC -c main.C
 
/* -brtl enable runtime linkage. */
$ xlC main.o -L. -ltest -brtl -bexpall -o main
```

gibfahn commented:
>@ezequielgarcia I think I saw something in the guide @richardlau pointed to:

As @richardlau mentioned, node uses gcc not xlc to compile C++, so I'm not sure that'll work.

ezequielgarcia commented:
@gibfahn I'm not sure that matters, as gcc acts as frontend to the linker. The linker flags should be the same, disregarding what compiler is used.

@refack Hm, could be. Maybe using runtime-linking (`-brtl`) does the trick. In any case, I feel this PR has gone too far already. I'd say we should merge it, given it's working on Linux, BSD and OSX. We can solve this later: either finding the right linker flags, or deciding AIX can't support this kind of games.

ezequielgarcia commented:
@gireeshpunathil sure!

gibfahn commented:
> I think we land this as is. @ezequielgarcia I can work with you to open up the test for AIX, if that is fine with you.

+1 to that

refack commented:
@jasnell `node-review` has flagged you as still "Requesting Changes". AFAICT your comments have been addressed. Can I consider your objections dismissed and land this?
![image](https://user-images.githubusercontent.com/96947/30166692-0ad92a5e-93b2-11e7-832b-e5559bb4df42.png)

jasnell commented:
I haven't yet done a thorough re-review, so I can't sign off but you can consider my initial review addressed and clear it.

mhdawson commented:
@gireeshpunathil thanks for offering to work with @ezequielgarcia to get the test working for AIX. Can you open a new issue to track that effort ?

and +1 to going ahead and landing

thefourtheye commented on doc/api/os.md:
```diff
@@ -1170,6 +1170,43 @@ The following error codes are specific to the Windows operating system:
+### dlopen Constants
+
+If available on the operating system, the following constants
+are exported in `os.constants.dlopen`. See dlopen(3) for detailed
```
Care to link http://man7.org/linux/man-pages/man3/dlopen.3.html?

ezequielgarcia replied:
Isn't it done automagically. I think we already discussed this.

thefourtheye replied:
My bad. Yes, tools/doc/html.js will take care of this.

thefourtheye commented on doc/api/process.md:
```diff
@@ -638,6 +638,48 @@ process's [`ChildProcess.disconnect()`][].
+
+* `module` {Object}
+* `filename` {string}
+* `flags` {integer}.
```
Nit: Dot at the end is not required.

thefourtheye replied:
Also, it would be better to specify the default value of this optional parameter, which is `kDefaultFlags`. Right now `crypto` module's `.md` has examples for this.

ezequielgarcia replied:
Hm, the default value is platform specific.

thefourtheye replied:
All the more important to document it I believe.

thefourtheye commented on doc/api/process.md:
```diff
@@ -638,6 +638,48 @@ process's [`ChildProcess.disconnect()`][].
+The `process.dlopen()` method allows to dynamically load shared
+objects. It is primarily used by `require()` to load
+C++ Addons, and should not be used directly, except in special
+cases. In other words, [`require()`][] should be prefered over
```
preferred

ezequielgarcia replied:
👍

thefourtheye commented on src/node.cc:
```diff
@@ -2498,36 +2502,96 @@ struct node_module* get_linked_module(const char* name) {
 }
 
-// DLOpen is process.dlopen(module, filename).
+struct DLib {
```
Should this be a `struct`? If we keep it as a `class` and accept the values of `filename_` and `errmsg_` in the constructor, then we don't have to expose all the members as `public` right?

ezequielgarcia replied:
Already discussed. See above.

thefourtheye replied:
I couldn't find the actual discussion corresponding to this. I know it is exhausting to search through it, but if you could find it, please point me to it. I'll try again when I get back home.

ezequielgarcia replied:
On https://github.com/nodejs/node/pull/12794#discussion_r124620884 it was suggested to do a class. So I did it and commented 
 https://github.com/nodejs/node/pull/12794#issuecomment-312984056. Later, https://github.com/nodejs/node/pull/12794#discussion_r128185744, where it was suggested to do a struct instead.

thefourtheye replied:
Sorry, I didn't check properly then.

thefourtheye commented on src/node.cc:
```diff
@@ -2498,36 +2502,96 @@ struct node_module* get_linked_module(const char* name) {
+    uv_dlclose(&lib_);
+  }
+#endif
+};
```
Isn't calling `Close` necessary in the destructor? I see that we are explicitly calling `Close` everywhere, but I thought cleaning up in destructor is the best practice.

ezequielgarcia replied:
Is this a blocker for landing?

bnoordhuis replied:
The handle should only be closed on error.  It needs to stay open on success, otherwise the shared object would be unloaded again.

ezequielgarcia replied:
We are not closing everywhere. The handle is only closed on error. See https://github.com/nodejs/node/pull/12794#discussion_r137767595

thefourtheye commented on test/addons/dlopen-ping-pong/test.js:
```diff
@@ -0,0 +1,27 @@
+
+if (common.isWindows) {
+  common.skip('dlopen global symbol loading is not supported on this os.');
+  return;
```
The `return`s are not necessary. `skip` will take care of it.

ezequielgarcia replied:
OK.

thefourtheye commented on test/addons/dlopen-ping-pong/test.js:
```diff
@@ -0,0 +1,27 @@
+
+const bindingPath = require.resolve(`./build/${common.buildType}/binding`);
+process.dlopen(module, bindingPath,
+               os.constants.dlopen.RTLD_NOW | os.constants.dlopen.RTLD_GLOBAL);
```
These variables are defined in an `if` macro guard. If they are not available in the current OS, then this test should be skipped, right?

ezequielgarcia replied:
I guess so.

thefourtheye commented on test/parallel/test-binding-constants.js:
```diff
@@ -9,7 +9,8 @@ assert.deepStrictEqual(
 assert.deepStrictEqual(
-  Object.keys(constants.os).sort(), ['UV_UDP_REUSEADDR', 'errno', 'signals']
+  Object.keys(constants.os).sort(), ['UV_UDP_REUSEADDR', 'dlopen', 'errno',
+                                     'signals']
```
Style Nit: The second array would look better in the second line I guess.

bnoordhuis commented on src/node.cc:
```diff
@@ -2498,36 +2502,96 @@ struct node_module* get_linked_module(const char* name) {
+
+  const char* GetError() {
+    return errmsg_.c_str();
+  }
```
Since these methods are simple wrappers and only used once, I'd simply inline them at the call site.

ezequielgarcia replied:
You mean like this `Local<String> errmsg = OneByteString(env->isolate(), dlib.errmsg_.c_str());` @bnoordhuis ?

bnoordhuis commented on src/node.cc:
```diff
@@ -2498,36 +2502,96 @@ struct node_module* get_linked_module(const char* name) {
+  void Close() {
+    uv_dlclose(&lib_);
+  }
+#endif
```
Can you change this to `#endif  // !__POSIX__`?  (Note the two spaces before the `//`.)

ezequielgarcia replied:
Done.

ezequielgarcia commented:
Just pushed another branch addressing @bnoordhuis and @thefourtheye comments.
Changes:
- s/prefered/preferred
- documented dlopen default flags.
- removed `return` after `common.skip` in tests.
- removed `GetError` and `Handle` wrappers from DLib struct, and instead access the members directly.
- added comment in #endif.

refack commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10000/
(ohh boy 5 figure job ID)

refack commented:
Landed in 5f22375922
(after _only_ 4 months and two prior PR attempts)
🍾

addaleax commented:
@refack thanks for taking care of this!
